### DISCUSSION TITLE
[question] postStart implementation

### DISCUSSION THREAD

**Discussion by gabrielcsapo** [facebook/docusaurus#5616] [Ideas]
Hello,

Is there work being done to implement this? We want to add local search functionality that doesn't require a full build in https://github.com/gabrielcsapo/docusaurus-plugin-search-local.

Thank you,
Gabriel Csapo

**Comment by slorber** [comment]
No work is being done on this but we could add it.

Can you detail the use-case?

When should such a lifecycle hook be called exactly and with which params/data?

II can think of 2 places:

```js
  devServer.listen(port, host, (err) => {
    // ...
  });
```

```js
compiler.hooks.done.tap('done', (stats) => {
  // ...
});
```

**Reply by gabrielcsapo** [reply]
These seem like the right places in my opinion. 

The plugin that I am writing basically needs to build a search index based on the final output of a build. It only needs to happen once we start the server and could be called after any iterative change. That might be CPU intensive so maybe having that hook be called as a separate thing would be better.

// called once when the dev server finishes building the first time
postStart() {}

// called after any rebuild finishes? 
postRebuild() {}

**Reply by gabrielcsapo** [reply]
In terms of the parameters, it should be the same as postBuild to keep the interface the same.

**Reply by gabrielcsapo** [reply]
thoughts on this @slorber ?

**Reply by slorber** [reply]
> These seem like the right places in my opinion.

Those are 2 different places, so we'll need to figure out which one of the 2 should trigger the lifecycle event 🤪 

After Webpack compilation looks better to me, similarly to `postBuild`. But the Webpack compilation only happens after a HTTP request, so maybe it can be a problem?

> `postRebuild`

Couldn't this be a boolean param in `postStart`?

> In terms of the parameters, it should be the same as postBuild to keep the interface the same.

Maybe then having `postCompile` after any webpack dev/prod compilation be a better lifecycle hook?

**Reply by gabrielcsapo** [reply]
I think postCompile would be a good hook, it is more verbose in what it actually means instead of postStart which is sort of confusing.

```
postCompile(rebuild: boolean)
```

**Reply by gabrielcsapo** [reply]
I created a PR with initial work that handles the postCompile hook case @slorber

**Reply by slorber** [reply]
Just linking, as it's not automatic on GH discussions: PR is https://github.com/facebook/docusaurus/pull/7737
